File: apps/api/app/models/contracts.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


AnalysisStatus = Literal["queued", "running", "completed", "failed"]
MarkerSeverity = Literal["low", "medium", "high"]
MarkerType = Literal[
    "strong_hook",
    "attention_drop",
    "deadspace_candidate",
    "high_rewatch_moment",
    "pacing_issue",
    "audio_energy_drop",
]
ConfidenceBand = Literal["low", "medium", "high"]
CutType = Literal["deadspace", "low_value"]
AudienceWorldStatus = Literal["hydrating", "ready", "partial", "unavailable"]


class VideoAsset(BaseModel):
    uploadId: str
    filename: str
    sourceUrl: str
    thumbnailUrl: str
    sourceStorageId: str | None = None
    thumbnailStorageId: str | None = None
    durationSec: float
    width: int
    height: int
    sizeBytes: int
    recordedAt: datetime | None = None
    fileModifiedAt: datetime | None = None


class UploadResponse(BaseModel):
    uploadId: str
    video: VideoAsset


class AnalyzeRequest(BaseModel):
    uploadId: str
    convexScanId: str | None = None
    syncToConvexScan: bool = True


class HemisphereHeatmap(BaseModel):
    left: list[float] = Field(min_length=64, max_length=64)
    right: list[float] = Field(min_length=64, max_length=64)


class BrainResponsePoint(BaseModel):
    stimulusTimeSec: float
    segmentStartSec: float
    segmentDurationSec: float
    globalActivation: float
    leftHemisphereActivation: float
    rightHemisphereActivation: float
    rollingVariance: float
    activationDelta: float
    spikeScore: float
    dropScore: float
    audioEnergy: float
    motionScore: float
    transcriptDensity: float
    sceneChange: bool
    silenceOverlap: bool
    hemisphereHeatmap: HemisphereHeatmap


class MeshInfo(BaseModel):
    space: str = "fsaverage5"
    subject: str = "average"
    lagCompensationSec: float = 5.0
    totalVertices: int


class BrainResponsePayload(BaseModel):
    timeSeries: list[BrainResponsePoint]
    meshInfo: MeshInfo


class BrainSignalSummary(BaseModel):
    averageActivation: float
    averageMotion: float
    averageAudioEnergy: float
    averageTranscriptDensity: float


class AudienceTimelinePoint(BaseModel):
    startSec: float
    endSec: float
    sentiment: float
    interest: float
    clarity: float
    trust: float
    shareIntent: float
    dropoffRisk: float
    primaryReaction: str
    note: str


class AudienceVoice(BaseModel):
    speaker: str
    handle: str
    role: str
    platform: str
    stance: Literal["positive", "negative"] = "positive"
    quote: str


class AudienceOutlook(BaseModel):
    headline: str
    summary: str
    likelyPraise: list[str]
    likelyPushback: list[str]
    timeline: list[AudienceTimelinePoint]
    roomVoices: list[AudienceVoice] = Field(default_factory=list)


class AudienceWorldPlatformBreakdown(BaseModel):
    platform: str
    volume: int
    engagement: int
    leaning: str
    dominantNarratives: list[str] = Field(default_factory=list)


class AudienceWorldCohort(BaseModel):
    id: str
    label: str
    size: int
    leaning: str
    proofThreshold: str
    keyConcerns: list[str] = Field(default_factory=list)
    liked: list[str] = Field(default_factory=list)
    blocked: list[str] = Field(default_factory=list)
    representativeAgentIds: list[int] = Field(default_factory=list)
    momentIds: list[str] = Field(default_factory=list)


class AudienceWorldComment(BaseModel):
    id: str
    agentId: int | None = None
    speaker: str
    handle: str
    role: str
    platform: str
    content: str
    createdAt: datetime | None = None
    likes: int = 0
    shares: int = 0


class AudienceWorldThread(BaseModel):
    id: str
    platform: str
    dominantStance: str
    engagement: int
    replyCount: int = 0
    participatingCohortIds: list[str] = Field(default_factory=list)
    rootPost: AudienceWorldComment
    replies: list[AudienceWorldComment] = Field(default_factory=list)


class AudienceWorldAgentStats(BaseModel):
    totalActions: int = 0
    redditActions: int = 0
    twitterActions: int = 0


class AudienceWorldAgent(BaseModel):
    id: int
    displayName: str
    handle: str
    role: str
    platforms: list[str] = Field(default_factory=list)
    bio: str | None = None
    stats: AudienceWorldAgentStats = Field(default_factory=AudienceWorldAgentStats)


class AudienceWorldInterview(BaseModel):
    agentId: int
    prompt: str
    response: str
    platform: str | None = None
    cached: bool = True


class AudienceWorldEvidenceMoment(BaseModel):
    windowId: str
    startSec: float
    endSec: float
    headline: str
    reason: str
    threadIds: list[str] = Field(default_factory=list)
    cohortIds: list[str] = Field(default_factory=list)
    agentIds: list[int] = Field(default_factory=list)


class AudienceWorldPayload(BaseModel):
    status: AudienceWorldStatus
    simulationId: str
    platformBreakdown: list[AudienceWorldPlatformBreakdown] = Field(default_factory=list)
    cohorts: list[AudienceWorldCohort] = Field(default_factory=list)
    threads: list[AudienceWorldThread] = Field(default_factory=list)
    agents: list[AudienceWorldAgent] = Field(default_factory=list)
    interviews: list[AudienceWorldInterview] = Field(default_factory=list)
    evidenceMoments: list[AudienceWorldEvidenceMoment] = Field(default_factory=list)



class Marker(BaseModel):
    t: float
    type: MarkerType
    severity: MarkerSeverity
    explanation: str
    suggestion: str


class DeadspaceCut(BaseModel):
    id: str
    type: CutType = "deadspace"
    start: float
    end: float
    reason: str
    defaultSelected: bool = True
    recommendedAction: str = "Remove this section or bridge it with a harder cut."


class ActionBoard(BaseModel):
    keep: list[str] = Field(default_factory=list)
    fixNow: list[str] = Field(default_factory=list)
    testNext: list[str] = Field(default_factory=list)
    exportPlan: list[str] = Field(default_factory=list)


class TimelineSegment(BaseModel):
    id: str
    type: MarkerType | CutType
    label: str
    start: float
    end: float
    severity: MarkerSeverity
    reason: str
    recommendedAction: str
    cutId: str | None = None


class ExportArtifact(BaseModel):
    exportId: str
    createdAt: datetime
    trimmedVideoUrl: str
    trimmedVideoStorageId: str | None = None
    selectedCutIds: list[str]
    removedSeconds: float
    trimmedDurationSec: float


class ScoreSet(BaseModel):
    hookScore: int
    pacingScore: int
    retentionEstimate: int
    viralPotential: int
    confidence: ConfidenceBand
    helpingFactors: list[str]
    hurtingFactors: list[str]


class AnalysisSummary(BaseModel):
    strengths: list[str]
    weaknesses: list[str]
    overallRecommendation: str


class ArtifactLinks(BaseModel):
    rawPredictionsUrl: str | None = None
    providerRawJsonUrl: str | None = None
    processedJsonUrl: str
    cutListJsonUrl: str
    eventsCsvUrl: str
    segmentsJsonUrl: str
    trimmedVideoUrl: str | None = None
    trimmedVideoStorageId: str | None = None


class Diagnostics(BaseModel):
    device: str
    modelRepo: str
    modelCommit: str
    transcriptWordCount: int
    sceneChangeCount: int
    deadspaceSeconds: float
    trimmedDurationSec: float | None = None
    warnings: list[str]


class AnalysisPayload(BaseModel):
    analysisId: str
    analysisMode: Literal["brain_scan", "read_the_room"] = "brain_scan"
    video: VideoAsset
    brainResponse: BrainResponsePayload
    audienceOutlook: AudienceOutlook | None = None
    audienceWorld: AudienceWorldPayload | None = None
    brainSummary: BrainSignalSummary | None = None
    markers: list[Marker]
    deadspaceCuts: list[DeadspaceCut] = Field(default_factory=list)
    lowValueCuts: list[DeadspaceCut] = Field(default_factory=list)
    cutPlan: list[DeadspaceCut] = Field(default_factory=list)
    actionBoard: ActionBoard = Field(default_factory=ActionBoard)
    timelineSegments: list[TimelineSegment] = Field(default_factory=list)
    exports: list[ExportArtifact] = Field(default_factory=list)
    scores: ScoreSet
    summary: AnalysisSummary
    artifacts: ArtifactLinks
    diagnostics: Diagnostics


class AnalysisResponse(BaseModel):
    analysisId: str
    status: AnalysisStatus
    createdAt: datetime
    updatedAt: datetime
    error: str | None = None
    payload: AnalysisPayload | None = None


class AudienceWorldResponse(BaseModel):
    analysisId: str
    world: AudienceWorldPayload


class AudienceWorldInterviewRequest(BaseModel):
    agentIds: list[int] = Field(min_length=1)
    prompt: str
    platform: str | None = None


class AudienceWorldInterviewResponse(BaseModel):
    analysisId: str
    prompt: str
    cached: bool
    interviews: list[AudienceWorldInterview] = Field(default_factory=list)


class HealthResponse(BaseModel):
    ok: bool
    analysisBackend: Literal["tribe", "gemini", "mirofish"]
    pythonVersion: str
    ffmpegAvailable: bool
    ffprobeAvailable: bool
    huggingFaceTokenPresent: bool
    geminiApiKeyPresent: bool
    selectedDevice: str
    modelStatus: Literal["unloaded", "loaded", "error"]
    modelRepo: str
    modelCommit: str
    blockers: list[str]
    notes: list[str]


class CompareRequest(BaseModel):
    analysisIdA: str
    analysisIdB: str


class TrimRequest(BaseModel):
    """Request to trim a video using a subset of detected deadspace cuts."""

    cutIndices: list[int] | None = None
    cutIds: list[str] | None = None


class TrimResponse(BaseModel):
    analysisId: str
    trimmedVideoUrl: str
    trimmedVideoStorageId: str | None = None
    originalDurationSec: float
    trimmedDurationSec: float
    removedSeconds: float
    appliedCuts: list[DeadspaceCut]


class CompareSlice(BaseModel):
    label: str
    winner: Literal["A", "B", "tie"]
    aScore: int
    bScore: int


class CompareResponse(BaseModel):
    analysisIdA: str
    analysisIdB: str
    winner: Literal["A", "B", "tie"]
    winnerReason: str
    recommendation: str
    summary: list[str]
    slices: list[CompareSlice]


EditorProjectStatus = Literal["drafting", "queued", "running", "completed", "failed"]
EditorDraftStatus = Literal["queued", "running", "completed", "failed"]
EditorDraftStage = Literal[
    "queued",
    "preparing_clips",
    "ordering_story",
    "rendering_video",
    "finalizing",
    "completed",
    "failed",
]
OrderingConfidence = Literal["low", "medium", "high"]
RepurposeResultStatus = Literal["queued", "running", "completed", "failed"]
RepurposeStage = Literal[
    "queued",
    "preparing_source",
    "planning_variants",
    "rendering_variants",
    "finalizing",
    "completed",
    "failed",
]
RepurposeDurationTarget = Literal["source", "short"]


class EditorClipDescriptor(BaseModel):
    clipId: str
    uploadId: str
    localUploadId: str
    filename: str


class EditorGenerateRequest(BaseModel):
    convexProjectId: str
    clips: list[EditorClipDescriptor] = Field(min_length=2)


class RepurposeGenerateRequest(BaseModel):
    convexProjectId: str
    sourceUploadId: str
    localUploadId: str
    filename: str


class EditorDraftExport(BaseModel):
    videoUrl: str
    videoStorageId: str | None = None
    durationSec: float


class OrderedDraftClip(BaseModel):
    clipId: str
    uploadId: str
    filename: str
    sourceOrder: int
    resolvedOrder: int
    rationale: str
    transcriptPreview: str
    summary: str
    speechCoverage: float
    removedSeconds: float
    trimmedDurationSec: float
    outputStartSec: float
    outputEndSec: float
    recordedAt: datetime | None = None
    fileModifiedAt: datetime | None = None
    warnings: list[str] = Field(default_factory=list)
    appliedCuts: list[DeadspaceCut] = Field(default_factory=list)


class EditorDraftPayload(BaseModel):
    export: EditorDraftExport
    storylineSummary: str
    orderingConfidence: OrderingConfidence
    orderedClips: list[OrderedDraftClip]
    warnings: list[str] = Field(default_factory=list)


class EditorDraftResponse(BaseModel):
    draftId: str
    projectId: str
    status: EditorDraftStatus
    stage: EditorDraftStage | None = None
    progressPercent: int | None = Field(default=None, ge=0, le=100)
    statusMessage: str | None = None
    createdAt: datetime
    updatedAt: datetime
    error: str | None = None
    payload: EditorDraftPayload | None = None


class RepurposeSourceSummary(BaseModel):
    sourceUploadId: str
    filename: str
    durationSec: float
    summary: str
    speechCoverage: float


class RepurposeSegmentSummary(BaseModel):
    segmentId: str
    startSec: float
    endSec: float
    transcriptPreview: str
    summary: str


class RepurposeVariant(BaseModel):
    variantId: str
    title: str
    angleSummary: str
    rationale: str
    durationTarget: RepurposeDurationTarget
    durationSec: float
    videoUrl: str
    videoStorageId: str | None = None
    segmentCount: int
    segments: list[RepurposeSegmentSummary] = Field(default_factory=list)


class RepurposeResultPayload(BaseModel):
    source: RepurposeSourceSummary
    summary: str
    variants: list[RepurposeVariant] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)


class RepurposeResultResponse(BaseModel):
    resultId: str
    projectId: str
    status: RepurposeResultStatus
    stage: RepurposeStage | None = None
    progressPercent: int | None = Field(default=None, ge=0, le=100)
    statusMessage: str | None = None
    createdAt: datetime
    updatedAt: datetime
    error: str | None = None
    payload: RepurposeResultPayload | None = None

File: apps/web/src/lib/contracts.ts
export type AnalysisStatus = "queued" | "running" | "completed" | "failed";
export type AnalysisMode = "brain_scan" | "read_the_room";
export type MarkerSeverity = "low" | "medium" | "high";
export type MarkerType =
  | "strong_hook"
  | "attention_drop"
  | "deadspace_candidate"
  | "high_rewatch_moment"
  | "pacing_issue"
  | "audio_energy_drop";
export type ConfidenceBand = "low" | "medium" | "high";
export type CutType = "deadspace" | "low_value";

export type HealthResponse = {
  ok: boolean;
  analysisBackend: "tribe" | "gemini" | "mirofish";
  pythonVersion: string;
  ffmpegAvailable: boolean;
  ffprobeAvailable: boolean;
  huggingFaceTokenPresent: boolean;
  geminiApiKeyPresent: boolean;
  selectedDevice: string;
  modelStatus: "unloaded" | "loaded" | "error";
  modelRepo: string;
  modelCommit: string;
  blockers: string[];
  notes: string[];
};

export type VideoAsset = {
  uploadId: string;
  filename: string;
  sourceUrl: string;
  thumbnailUrl: string;
  sourceStorageId?: string | null;
  thumbnailStorageId?: string | null;
  durationSec: number;
  width: number;
  height: number;
  sizeBytes: number;
  recordedAt?: string | null;
  fileModifiedAt?: string | null;
};

export type UploadResponse = {
  uploadId: string;
  video: VideoAsset;
};

export type HemisphereHeatmap = {
  left: number[];
  right: number[];
};

export type BrainResponsePoint = {
  stimulusTimeSec: number;
  segmentStartSec: number;
  segmentDurationSec: number;
  globalActivation: number;
  leftHemisphereActivation: number;
  rightHemisphereActivation: number;
  rollingVariance: number;
  activationDelta: number;
  spikeScore: number;
  dropScore: number;
  audioEnergy: number;
  motionScore: number;
  transcriptDensity: number;
  sceneChange: boolean;
  silenceOverlap: boolean;
  hemisphereHeatmap: HemisphereHeatmap;
};

export type BrainSignalSummary = {
  averageActivation: number;
  averageMotion: number;
  averageAudioEnergy: number;
  averageTranscriptDensity: number;
};

export type AudienceTimelinePoint = {
  startSec: number;
  endSec: number;
  sentiment: number;
  interest: number;
  clarity: number;
  trust: number;
  shareIntent: number;
  dropoffRisk: number;
  primaryReaction: string;
  note: string;
};

export type AudienceVoice = {
  speaker: string;
  handle: string;
  role: string;
  platform: string;
  stance?: "positive" | "negative";
  quote: string;
};

export type AudienceOutlook = {
  headline: string;
  summary: string;
  likelyPraise: string[];
  likelyPushback: string[];
  timeline: AudienceTimelinePoint[];
  roomVoices?: AudienceVoice[];
};

export type AudienceWorldPlatformBreakdown = {
  platform: string;
  volume: number;
  engagement: number;
  leaning: string;
  dominantNarratives: string[];
};

export type AudienceWorldCohort = {
  id: string;
  label: string;
  size: number;
  leaning: string;
  proofThreshold: string;
  keyConcerns: string[];
  liked: string[];
  blocked: string[];
  representativeAgentIds: number[];
  momentIds: string[];
};

export type AudienceWorldComment = {
  id: string;
  agentId?: number | null;
  speaker: string;
  handle: string;
  role: string;
  platform: string;
  content: string;
  createdAt?: string | null;
  likes: number;
  shares: number;
};

export type AudienceWorldThread = {
  id: string;
  platform: string;
  dominantStance: string;
  engagement: number;
  replyCount: number;
  participatingCohortIds: string[];
  rootPost: AudienceWorldComment;
  replies: AudienceWorldComment[];
};

export type AudienceWorldAgentStats = {
  totalActions: number;
  redditActions: number;
  twitterActions: number;
};

export type AudienceWorldAgent = {
  id: number;
  displayName: string;
  handle: string;
  role: string;
  platforms: string[];
  bio?: string | null;
  stats: AudienceWorldAgentStats;
};

export type AudienceWorldInterview = {
  agentId: number;
  prompt: string;
  response: string;
  platform?: string | null;
  cached: boolean;
};

export type AudienceWorldEvidenceMoment = {
  windowId: string;
  startSec: number;
  endSec: number;
  headline: string;
  reason: string;
  threadIds: string[];
  cohortIds: string[];
  agentIds: number[];
};

export type AudienceWorld = {
  status: "hydrating" | "ready" | "partial" | "unavailable";
  simulationId: string;
  platformBreakdown: AudienceWorldPlatformBreakdown[];
  cohorts: AudienceWorldCohort[];
  threads: AudienceWorldThread[];
  agents: AudienceWorldAgent[];
  interviews: AudienceWorldInterview[];
  evidenceMoments: AudienceWorldEvidenceMoment[];
};

export type Marker = {
  t: number;
  type: MarkerType;
  severity: MarkerSeverity;
  explanation: string;
  suggestion: string;
};

export type DeadspaceCut = {
  id: string;
  type: CutType;
  start: number;
  end: number;
  reason: string;
  defaultSelected: boolean;
  recommendedAction: string;
};

export type ActionBoard = {
  keep: string[];
  fixNow: string[];
  testNext: string[];
  exportPlan: string[];
};

export type TimelineSegment = {
  id: string;
  type: MarkerType | CutType;
  label: string;
  start: number;
  end: number;
  severity: MarkerSeverity;
  reason: string;
  recommendedAction: string;
  cutId: string | null;
};

export type ExportArtifact = {
  exportId: string;
  createdAt: string;
  trimmedVideoUrl: string;
  trimmedVideoStorageId?: string | null;
  selectedCutIds: string[];
  removedSeconds: number;
  trimmedDurationSec: number;
};

export type ScoreSet = {
  hookScore: number;
  pacingScore: number;
  retentionEstimate: number;
  viralPotential: number;
  confidence: ConfidenceBand;
  helpingFactors: string[];
  hurtingFactors: string[];
};

export type AnalysisPayload = {
  analysisId: string;
  analysisMode?: AnalysisMode;
  video: VideoAsset;
  brainResponse: {
    timeSeries: BrainResponsePoint[];
    meshInfo: {
      space: string;
      subject: string;
      lagCompensationSec: number;
      totalVertices: number;
    };
  };
  audienceOutlook?: AudienceOutlook | null;
  audienceWorld?: AudienceWorld | null;
  brainSummary?: BrainSignalSummary | null;
  markers: Marker[];
  deadspaceCuts: DeadspaceCut[];
  lowValueCuts: DeadspaceCut[];
  cutPlan: DeadspaceCut[];
  actionBoard: ActionBoard;
  timelineSegments: TimelineSegment[];
  exports: ExportArtifact[];
  scores: ScoreSet;
  summary: {
    strengths: string[];
    weaknesses: string[];
    overallRecommendation: string;
  };
  artifacts: {
    rawPredictionsUrl: string | null;
    providerRawJsonUrl: string | null;
    processedJsonUrl: string;
    cutListJsonUrl: string;
    eventsCsvUrl: string;
    segmentsJsonUrl: string;
    trimmedVideoUrl: string | null;
    trimmedVideoStorageId?: string | null;
  };
  diagnostics: {
    device: string;
    modelRepo: string;
    modelCommit: string;
    transcriptWordCount: number;
    sceneChangeCount: number;
    deadspaceSeconds: number;
    trimmedDurationSec: number | null;
    warnings: string[];
  };
};

export type AnalysisResponse = {
  analysisId: string;
  status: AnalysisStatus;
  createdAt: string;
  updatedAt: string;
  error: string | null;
  payload: AnalysisPayload | null;
};

export type CompareResponse = {
  analysisIdA: string;
  analysisIdB: string;
  winner: "A" | "B" | "tie";
  winnerReason: string;
  recommendation: string;
  summary: string[];
  slices: Array<{
    label: string;
    winner: "A" | "B" | "tie";
    aScore: number;
    bScore: number;
  }>;
};

export type AudienceWorldResponse = {
  analysisId: string;
  world: AudienceWorld;
};

export type AudienceWorldInterviewResponse = {
  analysisId: string;
  prompt: string;
  cached: boolean;
  interviews: AudienceWorldInterview[];
};

export type ScanType = "single" | "compare";

export type SavedCompareResult = {
  winner: "A" | "B" | "tie";
  winnerReason: string;
  recommendation: string;
  summary: string[];
  slices: Array<{
    label: string;
    winner: "A" | "B" | "tie";
    aScore: number;
    bScore: number;
  }>;
};

export type TrimRequest = {
  cutIndices?: number[] | null;
  cutIds?: string[] | null;
};

export type TrimResponse = {
  analysisId: string;
  trimmedVideoUrl: string;
  trimmedVideoStorageId?: string | null;
  originalDurationSec: number;
  trimmedDurationSec: number;
  removedSeconds: number;
  appliedCuts: DeadspaceCut[];
};

export type CurrentUser = {
  _id: string;
  email: string | null;
  name: string | null;
  image: string | null;
};

export type SavedScanSummary = {
  _id: string;
  scanType?: ScanType;
  title: string | null;
  filename: string;
  secondaryFilename: string | null;
  thumbnailUrl: string | null;
  secondaryThumbnailUrl: string | null;
  uploadId: string;
  secondaryUploadId: string | null;
  localUploadId: string | null;
  status: AnalysisStatus;
  localAnalysisId: string | null;
  secondaryLocalAnalysisId: string | null;
  compareResult: SavedCompareResult | null;
  viralPotential: number | null;
  hookScore: number | null;
  pacingScore: number | null;
  retentionEstimate: number | null;
  deadspaceSeconds: number | null;
  trimmedDurationSec: number | null;
  analysisUrl: string | null;
  overviewRecommendation: string | null;
  selectedCutIds: string[];
  latestExportUrl: string | null;
  lastExportedAt: number | null;
  errorMessage: string | null;
  createdAt: number;
  updatedAt: number;
};

export type SavedScanRecord = SavedScanSummary;

export type EditorProjectStatus = "drafting" | "queued" | "running" | "completed" | "failed";
export type EditorDraftStage =
  | "queued"
  | "preparing_clips"
  | "ordering_story"
  | "rendering_video"
  | "finalizing"
  | "completed"
  | "failed";

export type EditorProjectClip = {
  _id: string;
  uploadId: string;
  localUploadId: string | null;
  filename: string;
  durationSec: number | null;
  sourceOrder: number;
  createdAt: number;
};

export type EditorProjectSummary = {
  _id: string;
  title: string;
  status: EditorProjectStatus;
  clipCount: number;
  latestLocalDraftId: string | null;
  latestExportUrl: string | null;
  storylineSummary: string | null;
  orderingConfidence: ConfidenceBand | null;
  warningCount: number;
  errorMessage: string | null;
  createdAt: number;
  updatedAt: number;
};

export type EditorProjectDetail = EditorProjectSummary & {
  clips: EditorProjectClip[];
};

export type OrderedDraftClip = {
  clipId: string;
  uploadId: string;
  filename: string;
  sourceOrder: number;
  resolvedOrder: number;
  rationale: string;
  transcriptPreview: string;
  summary: string;
  speechCoverage: number;
  removedSeconds: number;
  trimmedDurationSec: number;
  outputStartSec: number;
  outputEndSec: number;
  recordedAt?: string | null;
  fileModifiedAt?: string | null;
  warnings: string[];
  appliedCuts: DeadspaceCut[];
};

export type EditorDraftPayload = {
  export: {
    videoUrl: string;
    videoStorageId?: string | null;
    durationSec: number;
  };
  storylineSummary: string;
  orderingConfidence: ConfidenceBand;
  orderedClips: OrderedDraftClip[];
  warnings: string[];
};

export type EditorDraftResponse = {
  draftId: string;
  projectId: string;
  status: "queued" | "running" | "completed" | "failed";
  stage: EditorDraftStage | null;
  progressPercent: number | null;
  statusMessage: string | null;
  createdAt: string;
  updatedAt: string;
  error: string | null;
  payload: EditorDraftPayload | null;
};

export type RepurposeProjectStatus =
  | "drafting"
  | "queued"
  | "running"
  | "completed"
  | "failed";

export type RepurposeProjectVariant = {
  _id: string;
  variantKey: string;
  title: string;
  angleSummary: string;
  durationTarget: "source" | "short";
  durationSec: number;
  exportUrl: string | null;
  position: number;
};

export type RepurposeProjectSummary = {
  _id: string;
  title: string;
  status: RepurposeProjectStatus;
  sourceUploadId: string | null;
  sourceFilename: string | null;
  sourceDurationSec: number | null;
  latestLocalResultId: string | null;
  variantCount: number;
  summary: string | null;
  errorMessage: string | null;
  createdAt: number;
  updatedAt: number;
};

export type RepurposeProjectDetail = RepurposeProjectSummary & {
  variants: RepurposeProjectVariant[];
};

export type RepurposeStage =
  | "queued"
  | "preparing_source"
  | "planning_variants"
  | "rendering_variants"
  | "finalizing"
  | "completed"
  | "failed";

export type RepurposeResultSegment = {
  segmentId: string;
  startSec: number;
  endSec: number;
  transcriptPreview: string;
  summary: string;
};

export type RepurposeResultVariant = {
  variantId: string;
  title: string;
  angleSummary: string;
  rationale: string;
  durationTarget: "source" | "short";
  durationSec: number;
  videoUrl: string;
  videoStorageId?: string | null;
  segmentCount: number;
  segments: RepurposeResultSegment[];
};

export type RepurposeResultPayload = {
  source: {
    sourceUploadId: string;
    filename: string;
    durationSec: number;
    summary: string;
    speechCoverage: number;
  };
  summary: string;
  variants: RepurposeResultVariant[];
  warnings: string[];
};

export type RepurposeResultResponse = {
  resultId: string;
  projectId: string;
  status: "queued" | "running" | "completed" | "failed";
  stage: RepurposeStage | null;
  progressPercent: number | null;
  statusMessage: string | null;
  createdAt: string;
  updatedAt: string;
  error: string | null;
  payload: RepurposeResultPayload | null;
};
